// @polybase/util/dist/algorithems/x25519-xsalsa20-poly1305.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.symmetricDecryptFromEncoding = exports.symmetricEncryptToEncoding = exports.asymmetricDecryptFromEncoding = exports.asymmetricEncryptToEncoding = exports.symmetricDecrypt = exports.symmetricEncrypt = exports.asymmetricDecrypt = exports.asymmetricEncrypt = exports.getPublicKey = exports.generateKeyPair = void 0;
var tweetnacl_1 = __importDefault(require("tweetnacl"));
var util_1 = require("../util");
/**
 * Generate public/private key pairs
 *
 * @returns public/private key pair
 */
function generateKeyPair() {
    var privateKey = tweetnacl_1.default.randomBytes(32);
    return {
        version: 'x25519-xsalsa20-poly1305/asymmetric',
        privateKey: privateKey,
        publicKey: tweetnacl_1.default.box.keyPair.fromSecretKey(privateKey).publicKey,
    };
}
exports.generateKeyPair = generateKeyPair;
/**
 * Generate 65-byte uncompressed public key from private key
 *
 * @returns public key
 */
function getPublicKey(privateKey) {
    return tweetnacl_1.default.box.keyPair.fromSecretKey(privateKey).publicKey;
}
exports.getPublicKey = getPublicKey;
/**
 * Asymmetric encrypt bytes
 *
 * @returns encrypted data
 */
function asymmetricEncrypt(publicKey, data) {
    return __awaiter(this, void 0, void 0, function () {
        var ephemeralKeyPair, nonce, pk, ciphertext;
        return __generator(this, function (_a) {
            ephemeralKeyPair = tweetnacl_1.default.box.keyPair();
            nonce = tweetnacl_1.default.randomBytes(tweetnacl_1.default.box.nonceLength);
            pk = (0, util_1.stripPublicKeyPrefix)(publicKey);
            ciphertext = tweetnacl_1.default.box(data, nonce, pk, ephemeralKeyPair.secretKey);
            return [2 /*return*/, {
                    version: 'x25519-xsalsa20-poly1305/asymmetric',
                    nonce: nonce,
                    ephemPublicKey: ephemeralKeyPair.publicKey,
                    ciphertext: ciphertext,
                }];
        });
    });
}
exports.asymmetricEncrypt = asymmetricEncrypt;
/**
 * Asymmetric decrypt bytes
 *
 * @returns decrypted bytes
 */
function asymmetricDecrypt(privateKey, encryptedData) {
    return __awaiter(this, void 0, void 0, function () {
        var version, nonce, ephemPublicKey, ciphertext, recieverEncryptionPrivateKey, decryptedMessage;
        return __generator(this, function (_a) {
            version = encryptedData.version, nonce = encryptedData.nonce, ephemPublicKey = encryptedData.ephemPublicKey, ciphertext = encryptedData.ciphertext;
            // For backwards compatability, allow base version
            (0, util_1.assert)(version === 'x25519-xsalsa20-poly1305' ||
                version === 'x25519-xsalsa20-poly1305/asymmetric', 'Invalid encryption version');
            recieverEncryptionPrivateKey = tweetnacl_1.default.box.keyPair.fromSecretKey(new Uint8Array(privateKey)).secretKey;
            decryptedMessage = tweetnacl_1.default.box.open(ciphertext, nonce, ephemPublicKey, recieverEncryptionPrivateKey);
            if (!decryptedMessage)
                throw new Error('Unable to decrypt');
            return [2 /*return*/, decryptedMessage];
        });
    });
}
exports.asymmetricDecrypt = asymmetricDecrypt;
/**
 * Symmetric encrypt bytes
 *
 * @returns encrypt data
 */
function symmetricEncrypt(privateKey, data) {
    return __awaiter(this, void 0, void 0, function () {
        var nonce, ciphertext;
        return __generator(this, function (_a) {
            nonce = tweetnacl_1.default.randomBytes(tweetnacl_1.default.box.nonceLength);
            ciphertext = tweetnacl_1.default.secretbox(data, nonce, privateKey);
            return [2 /*return*/, {
                    version: 'x25519-xsalsa20-poly1305/symmetric',
                    nonce: nonce,
                    ciphertext: ciphertext,
                }];
        });
    });
}
exports.symmetricEncrypt = symmetricEncrypt;
/**
 * Symmetric decrypt bytes
 *
 * @returns decrypted bytes
 */
function symmetricDecrypt(privateKey, encryptedData) {
    return __awaiter(this, void 0, void 0, function () {
        var version, ciphertext, nonce, decryptedMessage;
        return __generator(this, function (_a) {
            version = encryptedData.version, ciphertext = encryptedData.ciphertext, nonce = encryptedData.nonce;
            (0, util_1.assert)(version === 'x25519-xsalsa20-poly1305/symmetric', 'Invalid encryption version');
            decryptedMessage = tweetnacl_1.default.secretbox.open(ciphertext, nonce, privateKey);
            if (!decryptedMessage)
                throw new Error('Unable to decrypt');
            return [2 /*return*/, decryptedMessage];
        });
    });
}
exports.symmetricDecrypt = symmetricDecrypt;
/**
 * Asymmetric encrypt string data as a given encoding (hex/base64). Defaults to base64.
 *
 * @returns encrypted data as hex
 */
function asymmetricEncryptToEncoding(publicKey, data, encoding) {
    if (encoding === void 0) { encoding = 'base64'; }
    return __awaiter(this, void 0, void 0, function () {
        var e;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, asymmetricEncrypt(publicKey, (0, util_1.decodeFromString)(data, 'utf8'))];
                case 1:
                    e = _a.sent();
                    return [2 /*return*/, (0, util_1.stringifyEncryptedData)(e, encoding)];
            }
        });
    });
}
exports.asymmetricEncryptToEncoding = asymmetricEncryptToEncoding;
/**
 * Asymmetric decrypt data from given encoding (hex/base64). Defaults to base64.
 *
 * @returns decrypted string
 */
function asymmetricDecryptFromEncoding(privateKey, hex, encoding) {
    if (encoding === void 0) { encoding = 'base64'; }
    return __awaiter(this, void 0, void 0, function () {
        var e, res;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    e = (0, util_1.parseEncrypedData)(hex, encoding);
                    return [4 /*yield*/, asymmetricDecrypt(privateKey, e)];
                case 1:
                    res = _a.sent();
                    return [2 /*return*/, (0, util_1.encodeToString)(res, 'utf8')];
            }
        });
    });
}
exports.asymmetricDecryptFromEncoding = asymmetricDecryptFromEncoding;
/**
 * Symmetric encrypt string data as a given encoding (hex/base64). Defaults to base64.
 *
 * @returns encrypted data as encoding
 */
function symmetricEncryptToEncoding(publicKey, data, encoding) {
    if (encoding === void 0) { encoding = 'base64'; }
    return __awaiter(this, void 0, void 0, function () {
        var e;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, symmetricEncrypt(publicKey, (0, util_1.decodeFromString)(data, 'utf8'))];
                case 1:
                    e = _a.sent();
                    return [2 /*return*/, (0, util_1.stringifyEncryptedData)(e, encoding)];
            }
        });
    });
}
exports.symmetricEncryptToEncoding = symmetricEncryptToEncoding;
/**
 * Symmetric decrypt data from given encoding (hex/base64). Defaults to base64.
 *
 * @returns decrypted string
 */
function symmetricDecryptFromEncoding(privateKey, hex, encoding) {
    if (encoding === void 0) { encoding = 'base64'; }
    return __awaiter(this, void 0, void 0, function () {
        var e, res;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    e = (0, util_1.parseEncrypedData)(hex, encoding);
                    return [4 /*yield*/, symmetricDecrypt(privateKey, e)];
                case 1:
                    res = _a.sent();
                    return [2 /*return*/, (0, util_1.encodeToString)(res, 'utf8')];
            }
        });
    });
}
exports.symmetricDecryptFromEncoding = symmetricDecryptFromEncoding;
