// @polybase/client/dist/Query.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryResponse = exports.Query = exports.QueryWhereOperatorRelationMap = exports.QueryWhereOperatorMap = void 0;
var Record_1 = require("./Record");
var util_1 = require("./util");
exports.QueryWhereOperatorMap = {
    '>': '$gt',
    '<': '$lt',
    '>=': '$gte',
    '<=': '$lte',
    '==': '$eq',
};
exports.QueryWhereOperatorRelationMap = {
    $gt: '$gte',
    $gte: '$gt',
    $lt: '$lte',
    $lte: '$lt',
    $eq: '$eq',
};
var Query = /** @class */ (function () {
    function Query(collection, client, onQuerySnapshotRegister, onRecordSnapshotRegister) {
        var _this = this;
        this.sort = function (field, direction) {
            var q = _this.clone();
            if (!q.params.sort)
                q.params.sort = [];
            q.params.sort.push([field, direction !== null && direction !== void 0 ? direction : 'asc']);
            return q;
        };
        this.limit = function (limit) {
            var q = _this.clone();
            q.params.limit = limit;
            return q;
        };
        this.after = function (after) {
            _this.params.after = after;
            return _this;
        };
        this.before = function (before) {
            _this.params.before = before;
            return _this;
        };
        this.where = function (field, op, value) {
            var q = _this.clone();
            var referencedValue = value instanceof Record_1.CollectionRecord
                ? value.reference()
                : value;
            if (!q.params.where)
                q.params.where = {};
            var apiOp = exports.QueryWhereOperatorMap[op];
            q.params.where[field] = op === '=='
                ? referencedValue
                : (0, util_1.addKeyValue)(exports.QueryWhereOperatorMap[op], referencedValue, (0, util_1.removeKey)(exports.QueryWhereOperatorRelationMap[apiOp], q.params.where[field]));
            return q;
        };
        this.get = function () { return __awaiter(_this, void 0, void 0, function () {
            var ast, isReadPubliclyAccessible, sixtyMinutes, res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.collection.getAST()];
                    case 1:
                        ast = _a.sent();
                        return [4 /*yield*/, this.collection.isReadPubliclyAccessible()];
                    case 2:
                        isReadPubliclyAccessible = _a.sent();
                        sixtyMinutes = 60 * 60 * 1000;
                        return [4 /*yield*/, this.client.request(this.request())
                                .send(isReadPubliclyAccessible ? 'none' : 'required', sixtyMinutes)];
                    case 3:
                        res = _a.sent();
                        return [2 /*return*/, new QueryResponse(this.collection, this.client, this.onQuerySnapshotRegister, this.onRecordSnapshotRegister, res.data, ast)];
                }
            });
        }); };
        this.key = function () {
            return "query:".concat(_this.collection.id, "?").concat(JSON.stringify(_this.params));
        };
        this.onSnapshot = function (fn, errFn) {
            return _this.onQuerySnapshotRegister(_this, fn, errFn);
        };
        this.request = function () {
            return {
                url: "/collections/".concat(encodeURIComponent(_this.collection.id), "/records"),
                method: 'GET',
                params: _this.params,
            };
        };
        this.clone = function () {
            var q = new Query(_this.collection, _this.client, _this.onQuerySnapshotRegister, _this.onRecordSnapshotRegister);
            q.params = __assign(__assign({}, _this.params), { sort: _this.params.sort ? __spreadArray([], _this.params.sort, true) : undefined, where: _this.params.where ? __assign({}, _this.params.where) : undefined });
            return q;
        };
        this.params = {};
        this.collection = collection;
        this.client = client;
        this.onQuerySnapshotRegister = onQuerySnapshotRegister;
        this.onRecordSnapshotRegister = onRecordSnapshotRegister;
    }
    return Query;
}());
exports.Query = Query;
var QueryResponse = /** @class */ (function (_super) {
    __extends(QueryResponse, _super);
    function QueryResponse(collection, client, onQuerySnapshotRegister, onRecordSnapshotRegister, response, ast) {
        var _this = _super.call(this, collection, client, onQuerySnapshotRegister, onRecordSnapshotRegister) || this;
        _this.toJSON = function () {
            return {
                data: _this.data,
                cursor: _this.cursor,
            };
        };
        var data = response.data, cursor = response.cursor;
        _this.data = data.map(function (record) {
            return new Record_1.CollectionRecordResponse(collection.id, record, ast, _this.collection, client, onRecordSnapshotRegister);
        });
        _this.cursor = cursor;
        return _this;
    }
    QueryResponse.prototype.previous = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.before(this.cursor.before).get()];
            });
        });
    };
    QueryResponse.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.after(this.cursor.after).get()];
            });
        });
    };
    return QueryResponse;
}(Query));
exports.QueryResponse = QueryResponse;
